x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/ranges/range_editor.tsx

```diff
@@ -0,0 +1,502 @@
+                placeholder={FROM_PLACEHOLDER}
+                isInvalid={isValidNumber(from) && !isRangeWithin(tempRange)}
+              />
+            </EuiFlexItem>
+            <EuiFlexItem grow={false}>
+              <EuiIcon type="sortRight" color="subdued" />
+            </EuiFlexItem>
+            <EuiFlexItem>
+              <EuiFieldNumber
+                value={isFinite(to) ? Number(to) : ''}
+                onChange={({ target }) => {
+                  const newRange = {
+                    ...tempRange,
+                    to: target.value ? Number(target.value) : -Infinity,
+                  };
+                  setTempRange(newRange);
+                  saveRangeIfValid(newRange);
+                }}
+                prepend={
+                  <EuiToolTip content={ltTooltipContent}>
+                    <EuiText size="s">{ltPrependLabel}</EuiText>
+                  </EuiToolTip>
+                }
+                fullWidth={true}
+                compressed={true}
+                placeholder={TO_PLACEHOLDER}
+                isInvalid={isValidNumber(to) && !isRangeWithin(tempRange)}
+              />
+            </EuiFlexItem>
+          </EuiFlexGroup>
+        </EuiFormRow>
+        <EuiFormRow fullWidth>
+          <EuiFieldText
+            compressed
+            value={safeLabel}
+            prepend={i18n.translate('xpack.lens.indexPattern.ranges.customIntervalPopoverLabel', {
+              defaultMessage: 'Label',
+            })}
+            onChange={({ target }) => {
+              const newRange = {
+                ...tempRange,
+                label: target.value,
+              };
+              setTempRange(newRange);
+              saveRangeIfValid(newRange);
+            }}
+          />
+        </EuiFormRow>
+      </EuiForm>
+    </EuiPopover>
+  );
+};
+
+const AdvancedRangeEditor = ({
+  ranges,
+  setRanges,
+  onToggleEditor,
+}: {
+  ranges: RangeType[];
+  setRanges: (newRanges: RangeType[]) => void;
+  onToggleEditor: () => void;
+}) => {
+  // use a local state to store ids with range objects
+  const [localRanges, setLocalRanges] = useState<LocalRangeType[]>(() =>
+    ranges.map((range) => ({ ...range, id: generateId() }))
+  );
+  // we need to force the open state of the popover from the outside in some scenarios
+  // so we need an extra state here
+  const [isOpenByCreation, setIsOpenByCreation] = useState(false);
+  const lastIndex = localRanges.length - 1;
+
+  const updateRanges = useCallback(
+    (updatedRanges: LocalRangeType[]) => {
+      setRanges(updatedRanges.map(({ id, ...rest }) => ({ ...rest })));
+      setLocalRanges(updatedRanges);
+    },
+    [setRanges]
+  );
+
+  const updateRangesDebounced = React.useMemo(() => debounce(updateRanges, 256), [updateRanges]);
+
+  const addNewRange = () => {
+    updateRanges([
+      ...localRanges,
+      {
+        id: generateId(),
+        from: -Infinity,
+        to: Infinity,
+        label: '',
```

Comment:
In Visualize, ranges are added starting from the end of the previous range.

Reply:
I reproduced the Visualise behaviour here as shown:

<img width="406" alt="Screenshot 2020-09-09 at 15 04 00" src="https://user-images.githubusercontent.com/924948/92601935-b6e5d100-f2ad-11ea-9b34-a772529e0a9b.png">

I thought about the "next good value": but the assumption of the unity as "next good value" didn't convinced me. It is possible to do it though.